### PR TITLE
Support `--enable-load-relative` inside binstubs 

### LINKED ISSUE #2928: ruby setup.rb failed
I'm having a problem or would like to suggest a feature.

My current problem is Running `ruby setup.rb` failed
```
$ sudo ruby setup.rb
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
ERROR:  While executing gem ... (NameError)
    uninitialized constant Gem::Specification::OpenSSL
```

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [x] Other

Here are my current environment details:

```
$ gem env version
not available, installation failed
```
```
$ git describe --tags
v3.0.4-739-gf62157c3
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### LINKED ISSUE #5245: How to bypass RubyGems prompts in CI - bundler's executable "bundle" conflicts with ...
Using WIndows Ruby 3.1 on Actions.  When trying to perform:
```
gem install bundler -v "~> 2"
```

The following error is generated:

```
  ERROR:  Error installing bundler:
  	"bundle" from bundler conflicts with C:/hostedtoolcache/windows/Ruby/3.1.0/x64/bin/bundle
```

When running locally, RubyGems prompts with:

```
bundler's executable "bundle" conflicts with C:/rubies/Ruby31-x64/bin/bundle
Overwrite the executable? [yN]
```

What's the recommended way to bypass the prompt in CI?

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
# Description:

~`ruby setup.rb` was broken probably since #2782. Requiring `openssl` fixes it.~

~Closes #2928.~

When ruby has been compiled with the `--enable-load-relative` configuration, ruby core generates binstub files with some bash code on top, making `gem install` not detect them as rubygems binstubs.

Closes #5245.

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -18,4 +18,4 @@ jobs:
       - name: Install Dependencies
         run: gem install "rake:>=10.4.2" "minitest:>=5.4.3" --conservative --no-document
       - name: Run Test
-        run: rake test
+        run: rake TEST=test/rubygems/test_build_commands.rb
```
Comment (bronzdoc) [inline .github/workflows/macos.yml]: Why we not running all the tests in Mac OS?

File: lib/rubygems/specification.rb [hunk 2]
```diff
@@ -2462,6 +2462,7 @@ def to_ruby
       next if handled.include? attr_name
       current_value = self.send(attr_name)
       if current_value != default_value(attr_name) || self.class.required_attribute?(attr_name)
+        require "openssl"
```
Comment (bronzdoc) [inline lib/rubygems/specification.rb]: We were not requiring "openssl" before https://github.com/rubygems/rubygems/pull/2782 🤔

File: .github/workflows/macos.yml (hunk 1 above)
Comment (deivid-rodriguez) [reply .github/workflows/macos.yml]: Just to save some CI time for debugging. I'll restore all tests later.

File: lib/rubygems/specification.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply lib/rubygems/specification.rb]: No, but after #2782 we're using the `OpenSSL` constant, so we need to require `openssl`. At least that's what the error is telling us.

Comment (bronzdoc) [reply lib/rubygems/specification.rb]: Ahh you are right, I missed that.

**Comment by deivid-rodriguez** [comment]
Ok, so this PR surfaced an unrelated bug in rubygems installer.

On MacOS 2.6.3 ruby seems compiled with the `--enable-load-relative` configure option. That generates the following `bundler` binstub upon installation:

```ruby
#!/bin/sh
# -*- ruby -*-
_=_\
=begin
bindir="${0%/*}"
exec "$bindir/ruby" "-x" "$0" "$@"
=end
#!/usr/bin/env ruby
#
# This file was generated by RubyGems.
#
# The application 'bundler' is installed as part of a gem, and
# this file is here to facilitate running it.
#

require 'rubygems'
```

Rubygems installer does not support this kind of binstub and thinks it's not a bundler binstub.

This is a related ruby-core ticket: https://bugs.ruby-lang.org/issues/13997

**Comment by bronzdoc** [comment]
@deivid-rodriguez can this be closed now that https://github.com/rubygems/rubygems/pulls?q=is%3Apr+is%3Aclosed was merged?

**Comment by deivid-rodriguez** [comment]
I guess the test could still be useful, so I changed the title to reflect that new scope and will work on it.

**Comment by simi** [comment]
If I understand well, `ruby setup.rb` is there to install RubyGems without any dependencies (like `rake`). Anyway we do already test `rake install` on CI. Wouldn't be useful to unify `ruby setup.rb` and `rake install` and test that on CI? We can call `setup.rb` in `rake install` for example.

**Comment by deivid-rodriguez** [comment]
This PR started as the title title suggests but then deviated to fixing the issue manifested by a test failure.

So I'm going to rename the PR and will try to rebase it to see how it was going.

Regarding the `ruby setup.rb` test, I don't think it's necessary anymore, since as you point out we already test `rake install`. `rake install` ends up running `ruby setup.rb` under the hood. It'ss essentially the same thing, but includes building a `rubygems-update` package so it's more similar to reality.

**Comment by deivid-rodriguez** [comment]
I verified manually that this fixes #5245 so I will merge it.

**Comment by nobu** [comment]
`TestGem#test_looks_for_gemdeps_files_automatically_from_binstubs` and `TestGem#test_looks_for_gemdeps_files_automatically_from_binstubs_in_parent_dir` now fail because the installed directory doesn't contain `ruby` executable.

I think the symlink at the directory (or installing full set) could fix the failures, but it may not be a good way in the tests.

**Comment by deivid-rodriguez** [comment]
How do they fail? Everything seems green both here and in ruby-core, no?

**Comment by nobu** [comment]
https://github.com/nobu/ruby/runs/4826739682?check_suite_focus=true#step:16:132
>   /home/runner/work/ruby/ruby/build/tmp/test_rubygems_20220115-18636-37l1rc/gd-tmp/bin/foo: 6: exec: /home/runner/work/ruby/ruby/build/tmp/test_rubygems_20220115-18636-37l1rc/gd-tmp/bin/ruby: not found

It's green because there is no load-relative tests.

**Comment by deivid-rodriguez** [comment]
Oh I see, you're running the whole RubyGems suite against a Ruby with `--enable-load-relative`. I see, well only a couple of failures seems quite good :), I'll try it and see if there's an easy fix.

**Comment by NicholasBHubbard** [comment]
Opened #8135

**Pull request merged**